### ISSUE TITLE
update typed rosette to work with latest turnstile

### ISSUE THREAD

**Issue by bennn** [stchang/typed-rosette#36]
On racket version 7.0.0.8

And on typed-rosette commit: cc6b31cab090ef9a35d4700b02a9ab005473e4bc

Running `raco make typed/rosette/base-forms.rkt` gives this error:

```
typed/rosette/base-forms.rkt:209:43: x-: pattern variable cannot be used outside of a template
  after encountering unbound identifier (which is possibly the real problem):
   ⊢
  in: x-
```

EDIT: here's the report from the package server: http://pkg-build.racket-lang.org/server/built/fail/typed-rosette.txt

**Comment by AlexKnauth** [comment]
The `macrotypes/typecheck.rkt` file provides the `⊢` type assignment macro, but `turnstile` doesn't re-provide it. It only re-provides `macrotypes/typecheck-core.rkt`.

This appears to have changed in this commit of `turnstile`:
https://github.com/stchang/macrotypes/commit/6a7f0699d3d549f22c0d1a9a615ab30bcc689cbd

@stchang: What is the right thing to do? 
 (1) Should `turnstile` provide the `⊢` type assignment macro, or 
 (2) should all uses of the `⊢` macro be translated into uses of `assign-type` instead?

**Comment by AlexKnauth** [comment]
There are a lot of places where Typed Rosette uses the pattern:
```racket
(define-syntax- x (make-rename-transformer (assign-type #'y #'type)))
```
Which now breaks because `assign-type` now wraps it in the `erased` form for the stop-list.

Those should all be replaced with
```racket
(define-typed-variable-rename x ≫ y : type)
```
Is that right?

**Comment by stchang** [comment]
> What is the right thing to do?

This one:
> (2) should all uses of the ⊢ macro be translated into uses of assign-type instead?

Except, it should also specify `#:wrap? #f`.

And yes, `define-typed-variable-rename` was added to abstract these cases (`define-typed-variable` may be better in the cases that do `define`+`define-syntax`).

ps - Were you able to run any tests? I'm seeing syntax errors from travis.

**Comment by AlexKnauth** [comment]
I will change them to use `#:wrap? #f` and see if that improves things.

**Comment by stchang** [comment]
That wont get rid of all the errors, because I'm seeing some (Turnstile) syntax errors. I'm also looking into it.

**Comment by AlexKnauth** [comment]
I just pushed. The tests are still failing, but they're failing later, with errors more constrained to specific features. In particular the basic "rosette-tests.rkt" file is now compiling.

**Comment by AlexKnauth** [comment]
There's another bug that might be caused by the difference between Turnstile before the `erased` stop-list form, and Turnstile after. Some of the Typed Rosette forms use the `expand/ro` + `typeof` to do typechecking, and this `expand/ro` function doesn't include `erased` in the stop-list. 

For example an `(ann 1 : Int)` form would expand the `1` into `(erased '1) +: Nat`, and then wrap that in `erased` to get `(erased (erased '1) +: Nat) +: Int`. As long as the two properties are kept separate, this expression has the correct type: `Int`. 

However within `expand/ro`, it expands the `erased` away and merges the properties together with `cons`, mixing the "old" type with the "new" type, and resulting in the `ann` failing to override the old type.

At least that's what I suspect is happening. @michaelballantyne might know more about this. Since he's the one who created the `erased` stop-list strategy, he might be able to tell me how to use it properly with the `expand/ro` function.

**Comment by michaelballantyne** [comment]
If you adopt the `erased` stop-list strategy, it should no longer be necessary to specialize checking in typed rosette to stop on rosette core forms with something like `expand/ro`. My understanding was that was previously necessary because those Rosette forms use things like syntax parameters, so they need to be expanded in the usual order rather than inside-out, as the naive Turnstile expansion strategy would have done.

With `erased` wrapping all the type-erased syntax that typerule macros construct, the expansion of these forms will be properly delayed until all typechecking is complete without needing to add them to the stoplist specifically.

**Comment by stchang** [comment]
Yes, we knew that porting typed rosette to the new Turnstile (with stop lists) would be a larger effort, so I was putting it off temporarily. I was hoping to fix the current tests without the stop list first, but if you have some ideas for what else to fix, now would be a good time to look into it I guess.
